### PR TITLE
[ENH] Time Series Regression GridSearchCV

### PR DESCRIPTION
**Pull request by ksharma6**
#### Reference Issues/PRs

Related to issue #5655, and PRs #5986  #6075

#### What does this implement/fix? Explain your changes.
In working on the docstring for the MulitplexRegressor, there was not an Sktime implementation of Sklearn's GridSearchCV for regression module. This PR is an initial implementation.

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?
I wasn't sure what test to use to validate this implementation, please let me know which I can refer to to resolve any bugs. Thank you so much

#### Did you add any tests for the change?
No, please let me know if I need to

### CODE REVIEW

**Comment by yarnabrina** [comment]
Probably a dumb question: is it necessary to have separate grid/random/etc. search functionalities for forecasting/classification/regression/etc.? Can there not be a single implementation to reduce significant code duplication?

Or at least one base class and slight modification by inheritance per component?

**Comment by ksharma6** [comment]
hi @yarnabrina, that's not a dumb question at all.

@fkiraly had briefly mentioned implementing something similar to me not too long ago to help remove redundancies in the code base. I am pretty new to  sktime and am still learning the code base, happy to try implementing a base Cross validator and we can iterate over it? I'm also happy to wait on starting on something like this as well until I am a little less green with the code base, thoughts?

**Comment by fkiraly** [comment]
> Probably a dumb question: is it necessary to have separate grid/random/etc. search functionalities for forecasting/classification/regression/etc.? Can there not be a single implementation to reduce significant code duplication?

> Or at least one base class and slight modification by inheritance per component?

@yarnabrina, I have tried recently, and I think this is the right idea, but it is quite difficult to implement.

For this abstraction, one needs:

* an abstraction of task specific evaluator
* a structure that does not interfer with parallelization backends. This is very fiddly, as for example it is not without problems possible to dispatch to class methods

**Comment by ksharma6** [comment]
@fkiraly yes please :)

**Comment by yarnabrina** [comment]
@ksharma6 can you please add some tests? Given @fkiraly's explanation above, it makes sense to approve the testing of the current changes, but without tests it won't really do anything I suppose. May be you can add few simitests as it's present for classification/forecasting etc.?

(Also, please resolve the conflict.)

**Comment by ksharma6** [comment]
@fkiraly it looks like there isn't a test for `_tune.py` for classification (i don't see a `tests` folder, are the tests somehwere else?). Should I implement this in this PR or create it as a new issue and addres it in a seperate PR?

File: sktime/regression/model_selection/_tune.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+        from sktime.regression.distance_based import KNeighborsTimeSeriesRegressor
+
+        param1 = {
+            "estimator": KNeighborsTimeSeriesRegressor(distance="dtw"),
+            "param_grid": {"n_neighbors": [1, 3, 5]},
+        }
+
+        param2 = {
+            "estimator": KNeighborsTimeSeriesRegressor(distance="euclidean"),
+            "param_grid": {"algorithm": ["auto", "ball_tree", "kd_tree", "brute"]},
```
Comment (yarnabrina) [inline sktime/regression/model_selection/_tune.py:433]: It's a bit counter-intuitive to tune with `auto` option. I do not use time series regression myself. Will it make sense to remove `auto` from this grid?

File: sktime/regression/model_selection/_tune.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+            Each dict are parameters to construct an "interesting" test instance, i.e.,
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`.
+        """
+        from sklearn.metrics import r2_score
+
+        from sktime.regression.distance_based import KNeighborsTimeSeriesRegressor
+
+        param1 = {
+            "estimator": KNeighborsTimeSeriesRegressor(distance="dtw"),
```
Comment (yarnabrina) [inline sktime/regression/model_selection/_tune.py:427]: I'll suggest to consider using a different estimator just for the sake of slightly more coverage.

File: sktime/regression/model_selection/_tune.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+        "y_inner_mtype": ["numpy2D"],
+        "capability:multivariate": True,
+        "capability:multioutput": True,
+        "capability:unequal_length": True,
+        "capability:missing_values": True,
+        "capability:multithreading": True,
+        "capability:predict_proba": True,
+    }
+
+    def __init__(
```
Comment (yarnabrina) [inline sktime/regression/model_selection/_tune.py:247]: Have you seen this?

https://github.com/sktime/sktime/pull/6031

Will it make sense to add this support now only? I know @fkiraly suggested that this is a copy of classification counterpart so too many changes is out of scope, but this seems quite related to do in a different small PR soon afterwards.

File: sktime/regression/model_selection/_tune.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+        self,
+        estimator,
+        param_grid,
+        scoring=None,
+        n_jobs=None,
```
Comment (yarnabrina) [inline sktime/regression/model_selection/_tune.py:252]: Should we remove this in favour of backend parameters?

File: sktime/regression/model_selection/_tune.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+        fitted_param_names = [
+            "cv_results_",
+            "best_estimator_",
+            "best_score_",
+            "best_params_",
+            "best_index_",
+            "scorer_",
+            "n_splits_",
+            "refit_time_",
+            "multimetric_",
```
Comment (yarnabrina) [inline sktime/regression/model_selection/_tune.py:340]: Just curious: does it mean users can pass multiple metrics? We have a open PR for forecasting where more and more bugs come up over days.

https://github.com/sktime/sktime/pull/6030

File: sktime/regression/model_selection/_tune.py (hunk 1 above)
Comment (ksharma6) [reply sktime/regression/model_selection/_tune.py:433]: yes, good catch!

File: sktime/regression/model_selection/_tune.py (hunk 3 above)
Comment (ksharma6) [reply sktime/regression/model_selection/_tune.py:247]: @fkiraly do we want to add support from #6031 to this PR?

File: sktime/regression/model_selection/_tune.py (hunk 4 above)
Comment (ksharma6) [reply sktime/regression/model_selection/_tune.py:252]: Hi @yarnabrina I'm pretty new to sktime and open-source, where can I find the backend params?

File: sktime/regression/model_selection/_tune.py (hunk 5 above)
Comment (ksharma6) [reply sktime/regression/model_selection/_tune.py:340]: @yarnabrina yes, users should be able to pass multiple metrics. I did notice that there doesn't appear to be a test suite for the classifcation GridSearch which makes me wonder if we're possibly missing more bugs. 

I am in favor of removing multimetric_ for now, until there's a resolution for the forecaster, thoughts?

Comment (yarnabrina) [reply sktime/regression/model_selection/_tune.py:340]: If it already supports, then there is no need to remove. I was just surprised that this was not added in forecasting first.

File: sktime/regression/model_selection/_tune.py (hunk 4 above)
Comment (yarnabrina) [reply sktime/regression/model_selection/_tune.py:252]: See [this PR](https://github.com/sktime/sktime/pull/5268) and [this script](https://github.com/sktime/sktime/blob/main/sktime/utils/parallel.py).

This type of arguments were there in lots of places, but I thought it was planned to deprecate them in favour of `backend` (where you pass `loky` or `joblib` etc.) and `backend_params` (where you pass parallelisation configurations).

File: sktime/regression/model_selection/_tune.py [hunk 6]
```diff
@@ -53,7 +53,7 @@ class TSRGridSearchCV(_DelegatedRegressor):
 
     refit : bool, str, or callable, default=True
         Refit an estimator using the best found parameters on the whole
-        dataset. If ``False``, the ``predict`` and ``predict_probab`` will not work.
+        dataset. If ``False``, the ``predict`` will not work.
```
Comment (VascoSch92) [inline sktime/regression/model_selection/_tune.py:56]:
```suggestion
        dataset. If ``False``, then ``predict`` will not work.
```

**Comment by ksharma6** [comment]
> @fkiraly it looks like there isn't a test for `_tune.py` for classification (i don't see a `tests` folder, perhaps the tests somehwere else?). Should I implement this in this PR or create it as a new issue and addres it in a seperate PR?

Hi @fkiraly returning back to this conversation. Do tests exist somewhere for the GridSearchCV implementations for forecasting and classification?

Additionally I am stumped about what to do with the failed test. There seems to be a `TypeError: NoneType takes no arguments` which is cause by a `RuntimeWarning: invalid value encountered in divide` in this line here from sktime/transformations/panel/channel_selection:
 `line_vec_norm = line_vec / np.sqrt(np.sum(line_vec**2))`

Not sure if there is something about using the `ComposableTimeSeriesForestRegressor` for testing but please let me know  your thoughts when you have a moment.

**Comment by fkiraly** [comment]
@ksharma6, apologies for missing this.

> Do tests exist somewhere for the GridSearchCV implementations for forecasting and classification?

Yes, for forecasting, you have `forecasting.model_selection.tests.test_tune`.
For classification, there are no dedicated tests, only the general API tests from the `TestAllClassifiers`, `TestAllEstimators` classes, the same you get when applying `check_estimator`. This is, perhaps, a situation that should be improved.

**Comment by fkiraly** [comment]
The failure, I think, come from `TimeSeriesForestRegressor`, it seems like a bug of `TimeSeriesForestRegressor`.

The architecture of `TimeSeriesForestRegressor` is very unsafe, it inherits from a `sklearn` base class which should not be directly used imo. The base class must have changed, and its initialization now assumes that a parameter, `self.criterion` is set.

This should be reported as a bug.

For now, I would suggest that you use another regressor.

It is also strange why the tets suite does not pick up the issue with `TimeSeriesForestRegressor`. I would guess, either the issue is new, or it is specific to the parameter combination you are using in the tuning, and not covered by `get_test_params` of it. Perhaps that is something to explore based on the bug report.

**Comment by ksharma6** [comment]
> @ksharma6, apologies for missing this.
> 
> > Do tests exist somewhere for the GridSearchCV implementations for forecasting and classification?
> 
> Yes, for forecasting, you have `forecasting.model_selection.tests.test_tune`. For classification, there are no dedicated tests, only the general API tests from the `TestAllClassifiers`, `TestAllEstimators` classes, the same you get when applying `check_estimator`. This is, perhaps, a situation that should be improved.

No worries at all @fkiraly Thank you so much, I'll get started on this :)

**Comment by fkiraly** [comment]
> Would we want to do our own `TimeSeriesForestRegressor` implementation at some point?

Perhaps - at least we should clean the inheritance structure up to use only public interfaces.

**Comment by fkiraly** [comment]
let me know if you need a review or advice!

**Comment by ksharma6** [comment]
Still getting the following value error when running `check_estimator`:
`ValueError: Data seen by KNeighborsTimeSeriesRegressor instance has unequal length series, but this KNeighborsTimeSeriesRegressor instance cannot handle unequal length series. Calls with unequal length series may result in error or unreliable results.`

Let's see how changes perform against the test

**Comment by ksharma6** [comment]
Hi @fkiraly looks like I am getting the following error i mentioned earlier:
 `ValueError: Data seen by KNeighborsTimeSeriesRegressor instance has unequal length series, but this KNeighborsTimeSeriesRegressor instance cannot handle unequal length series. Calls with unequal length series may result in error or unreliable results.`

I didn't have this error before and I am not entirely sure what is causing it. Any ideas on what I may have done incorrectly?

**Review by fkiraly — CHANGES_REQUESTED** [review]
The failure is due to the example in `get_test_params` not being able to handle unequal length series, which is tested for the grid search.

A quick solution might be to do the same as for the classifier grid search, to pass distances or kernels that allow unequal length series.

There is a wider topic of cloning the tags appropriately, this is affecting both the classification and regression grid search and could be another work item - if you look at the forecasting grid search and how delegation or cloning of tags works there, it would be nice to have the same for classifiers, regressors. But I would consider that outside the scope of this PR.

**Comment by fkiraly** [comment]
tried the easy fix, let's see if it works

**Review by fkiraly — DISMISSED** [review]

**Pull request merged**